pkg/assetattributor/asset_inventory_api.go

```diff
@@ -0,0 +1,185 @@
+	// calls to asset-inventory-api's APIs for asset lookup by IP and hostname, wait for the calls
+	// to complete, and aggregrate the responses and errors onto buffered channels.
+	assetDetailsChan := make(chan domain.CloudAssetDetails, 2)
+	errChan := make(chan error, 2)
+	wg := &sync.WaitGroup{}
+
+	if asset.IP != "" {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			assetDetails, err := n.fetchAsset(ctx, resourcePathTypeIP, asset.IP, asset.LastScanned)
+			if err != nil {
+				errChan <- err
+				return
+			}
+			assetDetailsChan <- assetDetails[0]
+		}()
+	}
+
+	if asset.Hostname != "" {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			assetDetails, err := n.fetchAsset(ctx, resourcePathTypeHostname, asset.Hostname, asset.LastScanned)
+			if err != nil {
+				errChan <- err
+				return
+			}
+			assetDetailsChan <- assetDetails[0]
+		}()
+	}
+	wg.Wait()
+
+	// Once all goroutines have completed and all errors received, close the error channel and
+	// range over the buffered output, exiting on any fatal errors.
+	close(errChan)
+	outerErrs := []error{}
+	for e := range errChan {
+		switch e.(type) {
+		case httpMultipleAssetsFoundError:
+			return domain.NexposeAttributedAssetVulnerabilities{}, domain.AssetInventoryMultipleAssetsFoundError{
+				Inner:          e,
+				AssetID:        fmt.Sprintf("%d", asset.ID),
+				ScanTimestamp:  asset.LastScanned.Format(time.RFC3339Nano),
+				AssetInventory: assetInventoryAPIIdentifier,
+			}
+		case httpBadRequest:
+			return domain.NexposeAttributedAssetVulnerabilities{}, domain.AssetInventoryRequestError{
+				Inner:          e,
+				AssetID:        fmt.Sprintf("%d", asset.ID),
+				ScanTimestamp:  asset.LastScanned.Format(time.RFC3339Nano),
+				AssetInventory: assetInventoryAPIIdentifier,
+				Code:           http.StatusBadRequest,
+			}
+		}
+		outerErrs = append(outerErrs, e)
+	}
+
+	// Exit with an AssetNotFoundError error if both API calls returned non-fatal errors.
+	if len(outerErrs) == 2 {
+		return domain.NexposeAttributedAssetVulnerabilities{}, domain.AssetNotFoundError{
+			Inner:          combinedError{Errors: outerErrs},
+			AssetID:        fmt.Sprintf("%d", asset.ID),
+			ScanTimestamp:  asset.LastScanned.Format(time.RFC3339Nano),
+			AssetInventory: assetInventoryAPIIdentifier,
+		}
+	}
+
+	// Close the assetDetailsChan channel and append the buffered output to a slice.
+	close(assetDetailsChan)
+	assetDetails := []domain.CloudAssetDetails{}
+	for assetDetail := range assetDetailsChan {
+		assetDetails = append(assetDetails, assetDetail)
+	}
+
+	// Return the attributed asset with the first assetDetails returned by asset-inventory-api
+	return domain.NexposeAttributedAssetVulnerabilities{
+		NexposeAssetVulnerabilities: asset,
+		BusinessContext:             assetDetails[0],
+	}, nil
+}
+
+func (n *AssetInventoryAPIAttributor) fetchAsset(ctx context.Context, idType string, id string, ts time.Time) ([]domain.CloudAssetDetails, error) {
+	u, e := url.Parse(n.Host)
```

Comment:
Suggestion: Now that this value is being loaded through a settings component you have the option of parsing the URL in the component's New() method. This would allow you to return an error early in the startup process rather than hitting that same error here.

If you did this you would also be able to inject a *url.URL which would prevent you from needing to parse in this part of the code at all.

Reply:
I'll fix that, thanks!

Reply:
Fixed.